src/System.Private.Uri/src/System/UriHelper.cs

```diff
@@ -298,6 +298,37 @@ static void EnsureCapacity(char[]? dest, int destSize, int requiredSize)
             }
         }
 
+        internal static unsafe char[] UnescapeString(string input, int start, int end, char[] dest,
+            ref int destPosition, char rsvd1, char rsvd2, char rsvd3, UnescapeMode unescapeMode, UriParser? syntax,
+            bool isQuery)
+        {
+            fixed (char* pStr = input)
+            {
+                return UnescapeString(pStr, start, end, dest, ref destPosition, rsvd1, rsvd2, rsvd3, unescapeMode,
+                    syntax, isQuery);
+            }
+        }
+
+        internal static unsafe char[] UnescapeString(char* pStr, int start, int end, char[] dest, ref int destPosition,
+            char rsvd1, char rsvd2, char rsvd3, UnescapeMode unescapeMode, UriParser? syntax, bool isQuery)
+        {
+            ValueStringBuilder pooledArray = new ValueStringBuilder(dest.Length);
+            pooledArray.Append(dest.AsSpan(0, destPosition));
+
+            UnescapeString(pStr, start, end, ref pooledArray, rsvd1, rsvd2, rsvd3, unescapeMode,
+                    syntax, isQuery);
+            destPosition = pooledArray.Length;
+
+            if (destPosition > dest.Length)
+            {
+                dest = new char[destPosition];
+            }
+
+            pooledArray.AsSpan().TryCopyTo(dest);
```

Comment:
How about instead doing:
```C#
if (pooledArray.Length > dest.Length)
{
    dest = pooledArray.AsSpan().ToArray();
}
else
{
    pooledArray.AsSpan(destPosition).TryCopyTo(dest.AsSpan(destPosition));
}
```
so that if the destination array is already large enough, we don't need to re-copy the portion that's already there?  This also takes advantage of `Span<T>.ToArray()` which should be slightly faster than allocating and doing the copy yourself.

Reply:
Could you please explain how this change avoids re-copying of a dest's portion? In the beginning, all dest content is copied to VSB, then AsSpan(destPosition) is the same as Slice(0, destPosition) which means the VSB content is copied from its beginning.

Reply:
> AsSpan(destPosition) is the same as Slice(0, destPosition)

No, `AsSpan(destPosition)` is the same as `AsSpan(destPosition, length - destPosition)`.